### ISSUE TITLE
[RFC] Branching scheme changes

### ISSUE THREAD

**Issue by cjp256** [Bareflank/hypervisor#468]
Some repos are actively worked under "master", some "dev".  Some repos don't even have "dev" branches.

The current scheme is poorly documented, not consistently followed, and confusing for newcomers.

Proposal:

(1) Follow suit for nearly all git-based OSS projects where feature development occurs in "master" branch.
(2) When a stabilization/release occurs, create a release branch (e.g. stable-1.2).
(3) Tags can be made from the release branches (no change here from what's documented).

**Comment by rianquinn** [comment]
Once the v1.2 code is more stable, we will use master as normal. The problem was, when we were creating the new build system, nothing in master would compile which is miss-leading to people coming to the project, since most people are so used to just downloading master, and doing what they want. 

For this reason, if we intend to break the code, it's development will occur in a "dev" branch until the dev branch is mildly stable. Dev should not be used by anyone not involved in it's development, in which case, it should only effect a couple core developers. 

Another way of stating this is.... we are doing exactly the process you outline, which the exception that if we are going to break the repo, we do it in a different branch until it's stable again.

**Comment by cjp256** [comment]
Master should never be broken.  If it's broken, it doesn't belong in /bareflank/ and instead should reside in your personal fork until your feature is complete and ready to merge.

**Comment by connojd** [comment]
I don't know how many projects use master for feature development.  Feature development usually introduces instability and as you say master should never be broken so Im confused as to why we would want to develop features directly on master.

We already tag releases on master so I think that adding a branch for a release in addition to tags is overkill.

**Comment by brendank310** [comment]
I'm +1 for clarifying the branching. Starting to look at the project after a long hiatus, it's not clear where  efforts should be focused. It seems the dev branch is future v1.2 tag, but will it be merged back into master before that happens? I think ideally feature development occurs in a forked repo, where collaborators merge their contributions until the feature is working and ready for a PR into master. Forks with the heaviest development can always be highlighted in the main repo (frequent contributors forks may be of interest to various parties).

As to branching on release, this makes sense from the possibility of having tags that have micro versions. Any fixes that are deemed adequate post initial tag are included into stable-x.y, which allows for v.x.y.z tags.

**Comment by rianquinn** [comment]
-1. Here is a good example of a project that is really large, doing something similar. I can dig up more if it's needed. 
https://github.com/nlohmann/json

We could have worked v1.2 in my own fork, and then had everyone else working out of that fork for months, but I don't see how that changes anything. Either way, the code is not being developed in master.

Once again, I think we are debating something that is hopefully a one time thing. We did this because the build system broken everything, and took a couple of months to get working, which you cannot do to master. Normally, new features are done in your own fork after doing an RFC, and then submitted as a PR in a stable form, meaning master is always able to at least build. 

Also to answer your question, dev will be merged into master (really it just ends up deleting everything in the repo, and adding in the 5 files that are needed moving forward), and then the dev branches will be deleted. This will be done once all of the API and example repos are building properly, which will not occur for at least 4 more weeks.

**Comment by brendank310** [comment]
Here's a counter-example of another large project that does not have a development branches:
https://github.com/torvalds/linux

1.2 isn't a feature. The new build system was the feature required working out of a new branch. Doing that in a fork would have made more sense, and once it was ready (is it ready?) it should be merged into master. So I think the granularity of a feature branch was too low in this case, and things that were not build system related snuck into that feature branch. Also, it doesn't look like there will be a trivial merge from dev into master.

Whether or not @cjp256's proposals are accepted or not, feature branches should be much more focused than 'dev'.

**Comment by rianquinn** [comment]
The merge is actually really simple. Delete everything in master, and replace with the super-build files which is just like 5 or 6 files. 

Also, I can come up with a lot of examples that support both approaches. The point is, there is no one right way to do it, so our approach is just as valid as another. the new v1.2 build system is not complete, and should be in about 4 weeks if all goes well. And like I said before, we could have all worked on this in a fork (there were 4 of us working it), so for example, we could have all done this in my fork, but all that would have done is moved 4 people from /bareflank to /rianquinn. Either way, it was a massive undertaking, and where the code is being developed doesn't matter in the end. @cjp256 would simply be cloning from /rianquinn instead from /bareflank, and all of the same issues would still be coming up (v1.2 not being stable, and people using unsupported versions of Linux with v1.1).

**Comment by brendank310** [comment]
In doing so, you've broken the history of the repositories that are broken out. All original commits have been squashed in the breakout repositories from what I can tell.

**Comment by rianquinn** [comment]
yes, the history is gone, but the code should be viewed as new as it has changed substantially. Each repo has it's own history from the point they were created, and v1.1 history will still be in the hypervisor repo if it's needed. Likely because of this, v1.2 will end up being v2.0 when it's all said and done. 

The breakout was needed because of all of the different compilations that occur. Each one of these repos gets cross compiled by different parts of the system depending on how many API repos you include, so the original structure of the code was simply a no-go moving forward. As a result though, the build system is really simple, consisting of a set of nested super builds.

**Comment by rianquinn** [comment]
@cjp256 I'm going to close this, but will compromise as well. dev will stay in place while we finish up the new build system and cleanup. Once it is done and merged, all future changes that break master like this (if any) will be done in someone's fork, and all of the developers will move to this fork until its able to be merged.
